src/clj/rems/db/duo.clj

```diff
@@ -0,0 +1,144 @@
+            [rems.schema-base :refer [LocalizedString]]
+            [schema.coerce :as coerce]
+            [schema.core :as s]))
+
+;; The CSV format has English texts only
+(s/defschema DuoCodeCsv
+  {:id s/Str
+   (s/optional-key :shorthand) (s/maybe s/Str)
+   :label s/Str
+   :description s/Str})
+
+;; The DB format has localized texts
+(s/defschema DuoCodeDb
+  {:id s/Str
+   (s/optional-key :shorthand) (s/maybe s/Str)
+   :label LocalizedString
+   :description LocalizedString})
+
+(def ^:private coerce-DuoCodeCsv
+  (coerce/coercer! DuoCodeCsv coerce/string-coercion-matcher))
+
+(def ^:private coerce-DuoCodeDb
+  (coerce/coercer! DuoCodeDb coerce/string-coercion-matcher))
+
+(def ^:private validate-DuoCodeDb
+  (s/validator DuoCodeDb))
+
+(defn- rows->maps
+  "Transform CSV `rows` into maps with the keys from the first header row."
+  [rows]
+  (mapv zipmap
+        (->> (first rows) ; header
+             (map keyword)
+             repeat)
+        (rest rows)))
+
+(defn upsert-duo-code!
+  "Insert a new or update an existing DUO code into the database."
+  [duo]
+  (let [data (-> duo
+                 validate-DuoCodeDb
+                 (dissoc :id)
+                 (->> (remove-vals #(cond (string? %) (str/blank? %)
+                                          (coll? %) (empty? %)
+                                          :else %)))
+                 json/generate-string)]
+    (when (:id duo)
+      (db/upsert-duo-code! {:id (:id duo)
+                            :data data}))))
+
+(defn- read-zip-entries
+  "Read the zip-file entries from the `stream` and returns those matching `re`."
+  [stream re]
+  (when stream
+    (loop [files []]
+      (if-let [entry (.getNextEntry stream)]
+        (if (re-matches re (.getName entry))
+          (let [baos (java.io.ByteArrayOutputStream.)]
+            (io/copy stream baos)
+            (recur (conj files {:name (.getName entry)
+                                :bytes (.toByteArray baos)})))
+          (recur files))
+        files))))
+
+(defn fetch-duo-releases-latest
+  "Fetches the latest DUO codes from the GitHub repository.
+
+  Returns the unpacked, formatted result."
+  []
+  (let [release-response (http/get "https://api.github.com/repos/EBISPOT/DUO/releases/latest"
+                                   {:accept :json
+                                    :as :json})
+        release-body (:body release-response)
+        release-tag (:tag_name release-body)
+        zip-response (http/get (:zipball_url release-body) {:as :stream})
+        csv-bytes (-> zip-response :body (java.util.zip.ZipInputStream.) (read-zip-entries #".*/duo.csv$") first :bytes)]
+    (with-open [reader (io/reader csv-bytes)]
+      (let [codes (for [duo (->> reader csv/read-csv rows->maps)]
+                    (-> duo
+                        coerce-DuoCodeCsv
+                        (update-existing :label (fn [en] {:en en}))
+                        (update-existing :description (fn [en] {:en en}))))]
```

Comment:
these utilize english translation - do we have a plan for `:fi` and `:sv`? e.g. api update or something. or maybe configuration?

Reply:
Latest thinking from me is that we'll load the texts into the database and allow the owners to localize them in the DB with some tooling (if and only if the instances will support more than English). We don't need that tooling for the demo I think, as EGA / THL would be English only. The tooling can be somewhat manual as a first try (maybe with the API) or something in the UI too if we so decide.

We can't use the translation files because we check that every key exists in the code as well as the translations. This would be a problem a) before everything is translated b) because the codes themselves and their translation keys would perhaps be dynamic they can't be checked easily (or without code generation which is an option).

I didn't want to copy the English to the other languages. I'd rather make sure the translation utilities use the default language as fallback.

Reply:
yeah, let's go with the :en for demo. API might be easiest to implement but i don't know how users want to update the translations. most translations i've seen are a really asynchronous process and the results come in every now and then from separate translation service providers.

Reply:
Currently they can just update the translations file as they wish but I don't want to use that. In the future they would need to translate any localisations in the db table. They already do this for all the items, resources, ... that they create themselves, in the admin editors. I think it should be a rare occurrence that new DUO codes come or the phrasing changes. But if the codes change an operator must manually insert the new localisations into the db, or we can add a DUO tab with the capability.